### PR TITLE
BUG: .asfreq on resample on PeriodIndex/TimedeltaIndex are not

### LINKED ISSUE #12926: BUG: resample().asfreq() looses end period with TimedeltaIndex
```
In [16]: df=pd.DataFrame(data=[1,3], index=[dt.timedelta(), dt.timedelta(minutes=3)])
In [17]: df
Out[17]:
          0
00:00:00  1
00:03:00  3

In [18]: df.resample('1T').mean()
Out[18]:
            0
00:00:00  1.0
00:01:00  NaN
00:02:00  NaN
00:03:00  3.0

In [19]: df.resample('1T').asfreq()
Out[19]:
            0
00:00:00  1.0
00:01:00  NaN
00:02:00  NaN
```

The two examples should give the same result. 
Tried it with a DatetimeIndex, and there I don't see this problem.

cc @jreback

### PR DESCRIPTION
**Pull request by jreback**
closes #12926

### CODE REVIEW

**Comment by jreback** [comment]
a tad hacky on this soln, as seems ``DatetimeIndex`` does lots of manipulations inside the binner to fix fencepost issues, while the other indexes (Period/Timedelta) don't seem to need this.

cc @MaximilianR created a structure to support common testing for resampling (the ``Base`` test class)

File: pandas/tseries/resample.py [hunk 1]
```diff
@@ -764,6 +768,20 @@ class TimedeltaResampler(DatetimeIndexResampler):
     def _get_binner_for_time(self):
         return self.groupby._get_time_delta_bins(self.ax)
 
+    def _adjust_binner_for_upsample(self, binner):
+        """ adjust our binner when upsampling """
+        ax = self.ax
+
+        if is_subperiod(ax.freq, self.freq):
```
Comment (MaximilianR) [inline pandas/tseries/resample.py]: FYI this could super up

**Comment by MaximilianR** [comment]
Nice re the tests.

Tangential: I had a go deprecating the `how` keyword to make the Period stuff much simpler. I ran into issues with simple upsampling into super periods; I don't think that is possible without `DatetimeIndex` machinery atm. It's on my list though.

**Comment by jreback** [comment]
yep - I think we can still use much of the DatetimeIndexResampler machinery - just return the PeriodIndex at the end (and of course handle special cases)

File: pandas/tseries/resample.py (hunk 1 above)
Comment (jreback) [inline pandas/tseries/resample.py]: doesn't work. I think there is something going on w.r.t. how the values come back for DTI vs. TDI, meaning TDI always adds a stamp on (then removes it here). There is lots of internal logic, e.g. ``_adjust_bin_edges`` that is doing things.

**Pull request closed without merging**